### PR TITLE
Support storing the `FrameKind` when creating a `Frame`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

We want to be able to ask the `Frame` how it was created.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1199548034582004/1202366470755781/f) _(internal)_

## 🔍 What does this change?

- Add a `TaggedBox` implementation
- Add `FrameKind`
- Store `FrameKind` inside of the `TaggedBox`
- Add a method to return `FrameKind` from `Frame`
- Resolve TODOs in tests: Add missing tests for the `FrameKind`

## 📜 Does this require a change to the docs?

- The new code is documented

## 🐾 Next steps

- Change the output when printing `Report` depending on the `FrameKind`

## 🛡 What tests cover this?

- The unit tests were adjusted to test this
- `miri` will help ensure safety

### CODE REVIEW

File: packages/libs/error-stack/src/frame.rs [hunk 1]
```diff
@@ -14,6 +17,20 @@ use core::{
 use crate::provider::{self, Demand, Provider};
 use crate::Context;
 
+/// Representation of the [`Frame`] how it was created.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum FrameKind {
+    /// Frame was created with [`Report::new()`] or [`change_context()`].
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/src/frame.rs:23]:
```suggestion
    /// Frame was created through [`Report::new()`] or [`change_context()`].
```
Tiny nit

File: packages/libs/error-stack/src/frame.rs [hunk 2]
```diff
@@ -14,6 +17,20 @@ use core::{
 use crate::provider::{self, Demand, Provider};
 use crate::Context;
 
+/// Representation of the [`Frame`] how it was created.
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/src/frame.rs:20]:
```suggestion
/// Classification of the contents of a [`Frame`], determined by how it was created.
```

File: packages/libs/error-stack/src/frame.rs [hunk 3]
```diff
@@ -308,6 +344,88 @@ impl VTable {
...
+    /// so the last bit will *always* be `0`.
+    const MASK: usize = !(mem::align_of::<*const T>() - 1);
+
+    /// Creates a new tagged pointer with the tag
+    ///
+    /// # Panics
+    ///
+    /// if the tag is too large to be stored next to a pointer
+    pub fn new(frame: T, kind: FrameKind) -> Self {
+        // Will only fail on 8-bit platforms which Rust currently is not supporting
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/src/frame.rs:368]:
```suggestion
        // Will only fail on 8-bit platforms which Rust currently does not support
```

File: packages/libs/error-stack/src/frame.rs [hunk 4]
```diff
@@ -308,6 +344,88 @@ impl VTable {
...
+        let raw = Box::into_raw(Box::new(frame));
+
+        let tag = kind == FrameKind::Context;
+        // Store the tag in the last bit, due to alignment, this is 0 for 16-bit and higher
+        Self(raw as usize | usize::from(tag), PhantomData)
+    }
+
+    /// Returns the tag stored inside the pointer
+    pub const fn kind(&self) -> FrameKind {
+        if self.0 & 1 == 1 {
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/src/frame.rs:382]: Add a comment (similar to above) stating that we store the tag in the last bit with it being 1 if context and 0 if attachment.

File: packages/libs/error-stack/src/macros.rs [hunk 5]
```diff
@@ -279,7 +280,11 @@ mod tests {
             "additional message",
             "Context A"
         ]);
-        // TODO: check `err.frames().next().kind() == FrameKind::Context`
+        assert_eq!(kinds(&err), [
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/src/macros.rs:283]: Can we also test one or two with multiple contexts and attachments, just to make sure?

Comment (TimDiekmann) [reply packages/libs/error-stack/src/macros.rs]: Added a new test:
```rust
#[test]
fn kinds() {
    use FrameKind::{Attachment, Context};

    let report = Report::new(ContextA);
    let report = report.attach("A1");
    let report = report.attach("A2");
    let report = report.change_context(ContextB);
    let report = report.attach("B1");
    let report = report.attach("B2");

    assert_eq!(frame_kinds(&report), [
        Attachment, Attachment, Context, Attachment, Attachment, Context
    ]);
    assert_eq!(messages(&report), [
        "B2",
        "B1",
        "Context B",
        "A2",
        "A1",
        "Context A"
    ]);
}
```

**Pull request merged**